library/psa_crypto.c

```diff
@@ -4388,11 +4388,67 @@ static psa_status_t psa_key_derivation_hkdf_read( psa_hkdf_key_derivation_t *hkd
     size_t hmac_output_length;
     psa_status_t status;
     const uint8_t last_block = PSA_ALG_IS_HKDF_EXTRACT( kdf_alg ) ? 0 : 0xff;
+    size_t data_length;
 
-    if( hkdf->state < HKDF_STATE_KEYED ||
-        ( ! hkdf->info_set && ! PSA_ALG_IS_HKDF_EXTRACT( kdf_alg ) ) )
-        return( PSA_ERROR_BAD_STATE );
-    hkdf->state = HKDF_STATE_OUTPUT;
+    if ( hkdf->inputs.state != HKDF_STATE_OUTPUT )
```

Comment:
This whole block is part of the input gathering stage, and should be in its own function. Furthermore, I think it's overly complex to call this from an algorithm-specific function: the code to finish processing the inputs and go from the input stage to the output stage should be called directly from `output_bytes`. The logic in `output_bytes` is:

1. If the operation is still at the input stage, call `psa_crypto_key_derivation_complete_inputs` (new function name).
2. Check the capacity.
3. Dispatch to the appropriate `read` function.

`psa_crypto_key_derivation_complete_inputs` checks that all required inputs are present and copies data from the `inputs` side of the union to the `running_context` part. Later it will dispatch to a driver if there is one.

Reply:
Good point. I will move this part to separate function (`psa_crypto_key_derivation_complete_inputs`) that will be called from `output_bytes()` to finish gathering stage.

At least in the current proposition `inputs` are already part of the `running_context` and initialized so no copy operation is required.